src/centralized.rs

```diff
@@ -0,0 +1,499 @@
+        });
+        result
+    }
+
+    /// Indexed proving algorithm, returns an empty proof if no suitable
+    /// candidate is found within the setup.b steps.
+    fn prove_index(setup: &Setup, set: &[Data], v: usize) -> (usize, Option<Proof>) {
+        let mut bins: Vec<Vec<Data>> = vec![];
+        for _ in 0..setup.n_p {
+            bins.push(vec![]);
+        }
+        for &s in set.iter() {
+            bins[Proof::h0(setup, v, s)].push(s);
+        }
+        let nb_steps = Arc::new(AtomicUsize::new(0));
+        for t in 0..setup.d {
+            if nb_steps.load(atomic::Ordering::Relaxed) == setup.b {
+                return (0, None);
+            }
+            nb_steps.fetch_add(1, atomic::Ordering::Relaxed);
+            let round = Round::new(v, t + 1, setup.n_p);
+            let res = Proof::dfs(setup, &bins, &round, nb_steps.clone());
+            if res.is_some() {
+                return (nb_steps.load(atomic::Ordering::Relaxed), res);
+            }
+        }
+        (nb_steps.load(atomic::Ordering::Relaxed), None)
+    }
+
+    /// Alba's proving algorithm, based on a depth-first search algorithm.
+    /// Calls up to setup.r times the prove_index function and returns an empty
+    /// proof if no suitable candidate is found.
+    pub fn prove(setup: &Setup, set: &[Data]) -> Self {
+        for v in 0..setup.r {
+            if let (_, Some(proof)) = Proof::prove_index(setup, set, v + 1) {
+                return proof;
+            }
+        }
+        Proof::new()
+    }
+
+    /// Alba's proving algorithm used for benchmarking, returning a proof as
+    /// well as the number of  steps ran to find it.
+    pub fn bench(setup: &Setup, set: &[Data]) -> (usize, usize, Self) {
+        let mut nb_steps = 0;
+        for v in 0..setup.r {
+            let (steps, opt) = Proof::prove_index(setup, set, v + 1);
+            nb_steps += steps;
+            if let Some(proof) = opt {
+                return (nb_steps, proof.r, proof);
+            }
+        }
+        (nb_steps, setup.r, Proof::new())
+    }
+
+    /// Alba's verification algorithm, follows proving algorithm by running the
+    /// same depth-first search algorithm.
+    pub fn verify(setup: &Setup, proof: Proof) -> bool {
+        if proof.r == 0
+            || proof.d == 0
+            || proof.d > setup.d
+            || proof.r > setup.r
+            || proof.items.len() != setup.u
+        {
+            return false;
+        }
+        let r0 = Round::new(proof.r, proof.d, setup.n_p);
+        let (b, round) = proof.items.iter().fold((true, r0), |(b, r), &s| {
+            (
+                b && r.h_usize == Proof::h0(setup, proof.r, s),
+                Round::update(&r, s),
+            )
+        });
+        b && Proof::h2(setup, &round)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use rand_chacha::ChaCha20Rng;
+    use rand_core::{RngCore, SeedableRng};
+
+    #[test]
+    fn test_params() {
```

Comment:
This function doesn't seem to test anything -- there are no `assert`s. Why do we have it?

Reply:
Removed it.
Pyrrhos had sent a summary of the paper for the parameter choice and it was to check whether this corresponded with the table borns ( < lamba, > lambda^2, ...)